Cython/Compiler/Parsing.py

```diff
@@ -3077,13 +3087,20 @@ def p_with_gil(s):
     else:
         return 0
 
-def p_exception_value_clause(s):
+def p_exception_value_clause(s, ctx):
     exc_val = None
```

Comment:
I started looking at the tristate thing as a follow-up, but it looks more complicated to me than previously thought. In particular, `exc_val` can be populated even if `exc_check='+'`, so I don't see a way to get rid of the tri-state without introducing another boolean/variable, which is likely not worth it.

In the meantime, perhaps the docstring I started writing might be useful for later readers of this functions (because for someone not knee-deep in cython entrails, I'd say it's not obvious 😅)

```suggestion
def p_exception_value_clause(s, ctx):
    """
    Parse exception value clause.

    Maps clauses to exc_check / exc_value as follows:
     ___________________________________________________
    |                           |           |           |
    | Clause                    | exc_check | exc_value |
    |___________________________|___________|___________|
    |                           |           |           |
    | <nothing> (default func.) | True      | None      |
    | <nothing> (cdef extern)   | False     | None      |
    | noexcept                  | False     | None      |
    | except <val>              | False     | <val>     |
    | except? <val>             | True      | <val>     |
    | except +                  | '+'       | None      |
    | except +*                 | '+'       | '*'       |
    | except +<PyErr>           | '+'       | <PyErr>   |
    |___________________________|___________|___________|
    """
    exc_val = None
```